jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ByteChannelContentSource.java

```diff
@@ -0,0 +1,304 @@
+            }
+
+            try
+            {
+                ByteBuffer byteBuffer = _buffer.getByteBuffer();
+                BufferUtil.clearToFill(byteBuffer);
+                if (_length >= 0)
+                    byteBuffer.limit((int)Math.min(_buffer.capacity(), _length - _totalRead));
+                int read = _byteChannel.read(byteBuffer);
+                BufferUtil.flipToFlush(byteBuffer, 0);
+                if (read == 0)
+                    return null;
+                if (read > 0)
+                {
+                    _totalRead += read;
+                    if (_length < 0 || _totalRead < _length)
+                    {
+                        _buffer.retain();
+                        return Content.Chunk.asChunk(byteBuffer, false, _buffer);
+                    }
+
+                    _terminal = Content.Chunk.EOF;
+                    IO.close(_byteChannel);
+                    Content.Chunk last = Content.Chunk.asChunk(byteBuffer, true, _buffer);
+                    _buffer = null;
+                    return last;
+                }
+                _buffer.release();
+                _buffer = null;
+                _terminal = Content.Chunk.EOF;
+                IO.close(_byteChannel);
+            }
+            catch (Throwable t)
+            {
+                _terminal = Content.Chunk.from(t, true);
+            }
+        }
+        return _terminal;
+    }
+
+    @Override
+    public void fail(Throwable failure)
+    {
+        try (AutoLock ignored = lock.lock())
+        {
+            if (_terminal == null)
+                _terminal = Content.Chunk.from(failure, true);
+            else
+                ExceptionUtil.addSuppressedIfNotAssociated(_terminal.getFailure(), failure);
+            IO.close(_byteChannel);
+            if (_buffer != null)
+            {
+                _buffer.release();
+                _buffer = null;
+            }
+        }
+    }
+
+    @Override
+    public long getLength()
+    {
+        return _length;
+    }
+
+    @Override
+    public boolean rewind()
+    {
+        try (AutoLock ignored = lock.lock())
+        {
+            lockedCheckOpen();
+            if (_terminal != null || _byteChannel == null || !_byteChannel.isOpen())
+                return false;
+
+            if (_offset >= 0 && _byteChannel instanceof SeekableByteChannel seekableByteChannel)
+            {
+                try
+                {
+                    seekableByteChannel.position(_offset);
+                    _totalRead = 0;
+                    return true;
+                }
+                catch (Throwable t)
+                {
+                    _terminal = Content.Chunk.from(t, true);
+                }
+            }
+            return false;
+        }
+    }
+
+    /**
+     * A {@link ByteChannelContentSource} for a {@link Path}
+     */
+    public static class PathContentSource extends ByteChannelContentSource
```

Comment:
Should not be an inner class, but rather be retrofitted into the existing top-level `PathContentSource`.

Reply:
That can't be done without changes in behaviour.  So will implement here for now, but as deprecated.  In 12.1 the PathContentSource can be changed to take this implementation directly